JMS Serializer library and bundle upgrade

dbalabka opened the pull request:
## Goal
Make compatible with new JMS Serializer release.

## Plan
We are going to upgrade JMS Serializer and check the state of a library by running existing test.

**TODO**
- [x] Set new versions of JMS Serializer and related libraries
- [x] Make tests work with new JMS Serializer
- [x] Cover more with tests to catch rest of the issues

dbalabka committed:
Update composer.json with new Symfony and JMS serializer

dbalabka committed:
Trigger build

dbalabka committed:
Fixed methods overrides
Upgrade dependency till 7.1 because of nullable types in methods declarations

dbalabka committed:
Fixed method override

dbalabka committed:
Deprecate old version of JMS because it can't be compatible with JMS 2.0

dbalabka committed:
Deprecate PHP version 7.1 because it is not compatible with JMS 2.0. Set minimal PHP version 7.2

dbalabka committed:
Deprecate PHP version 7.1 because it is not compatible with JMS 2.0. Set minimal PHP version 7.2

dbalabka committed:
Upgraded JMS Serializer till 2.1 and JMS Serializer Bundle till 3.0.
Upgraded PHPUnit till 7.0
Added support of PHP 7.2.
Removed support of yml and xml in DateHandler
Fixed some issues in HTTP converter tests

oleg-andreyev commented on Serializer/Handler/ArrayCollectionHandler.php:
```diff
@@ -9,21 +9,36 @@
+    /** @var BaseArrayCollectionHandler  */
+    private $arrayCollectionHandler;
+
+    public function __construct(BaseArrayCollectionHandler $arrayCollectionHandler)
```
there is no needs to decorate the original class, need to add `tags` to already defined `ArrayCollectionHandler`
and this override could be removed

oleg-andreyev commented on Serializer/Handler/DateHandler.php:
```diff
@@ -9,33 +9,106 @@
+
+    public function __construct(BaseDateHandler $dateHandler)
+    {
+        $this->dateHandler = $dateHandler;
```
there is no needs to decorate the original class, need to add `tags` to already defined `DateHandler`
and this override could be removed

oleg-andreyev commented on composer.json:
```diff
@@ -15,20 +15,20 @@
-        "symfony/expression-language": "~2.3|~3.0",
-        "jms/serializer-bundle": "~0.13|~1.0|~2.0",
-        "pagerfanta/pagerfanta": "~2.0",
+        "symfony/yaml": "^2.3|^3.4",
```
consider adding symfony 4

dbalabka committed:
Bumped supported Symfony version

dbalabka committed:
Update travis to run multiple Symfony versions. Remove Make file

dbalabka committed:
Implemented functional tests and migrated existing to use WebTestCase instead JsonApi builder (fixes #8)
Fixed issues with services configuration.
Removed cache configuration and rely on original jms_serializer.metadata.file_cache.dir config value.
Documentation updates.

dbalabka commented:
@oleg-andreyev please review

dbalabka committed:
Fixing issue with symfony/validation package installation

dbalabka commented:
@oleg-andreyev please ask me to rebase when you will finish review

dbalabka committed:
Added reference to LICENSE file.
Added 1.0 deprecation notice.

oleg-andreyev commented on Resources/config/services.yml:
```diff
@@ -107,3 +101,17 @@ services:
+        arguments:
+            - '@jms_serializer.datetime_handler'
+        tags:
+            - { name: jms_serializer.subscribing_handler }
```
Do we really need a separate class which does not even adds extra logic (as decorators should), I'd add a CompilerPass which will register subject `DateHandler` with `json:api` tag (review how subscribing_handler are handled to get the clue)

dbalabka replied:
@oleg-andreyev it seems that `getSubscribingMethods` is the only way how to declare format and there no other option to declare `DateHandler` for `json:api`

oleg-andreyev commented on Resources/config/services.yml:
```diff
@@ -107,3 +101,17 @@ services:
+            - { name: jms_serializer.subscribing_handler }
+
+    json_api.array_collection_handler:
+        class: Mango\Bundle\JsonApiBundle\Serializer\Handler\ArrayCollectionHandler
```
Same as above

dbalabka replied:
@oleg-andreyev it seems that `getSubscribingMethods` is the only way how to declare format and there no other option to declare `DateHandler` for `json:api`

oleg-andreyev commented on Serializer/Visitor/Factory/JsonApiSerializationVisitorFactory.php:
```diff
@@ -0,0 +1,59 @@
+     *
+     * @return JsonApiSerializationVisitorFactory
+     */
+    public function setShowVersionInfo($showVersionInfo)
```
I'd made `showVersionInfo` immutable

dbalabka replied:
There no usages of this method so I have removed it

oleg-andreyev commented on Tests/Functional/App/AnnotationOnly/bundles.php:
```diff
@@ -0,0 +1,11 @@
+<?php
+/*
+ * This file is part of the OpCart software.
```
fix this

oleg-andreyev commented on Tests/Functional/App/AppKernel.php:
```diff
@@ -0,0 +1,92 @@
+<?php
+
+/*
+ * This file is part of the FOSElasticaBundle package.
```
fix this

oleg-andreyev commented on Tests/Functional/App/AppKernel.php:
```diff
@@ -0,0 +1,92 @@
+use Symfony\Component\HttpKernel\Kernel;
+
+/**
+ * Taken from https://github.com/FriendsOfSymfony/FOSElasticaBundle/blob/master/tests/Functional/app/AppKernel.php
```
afaik original it's taken from https://github.com/symfony/symfony/blob/master/src/Symfony/Bundle/SecurityBundle/Tests/Functional/app/AppKernel.php#L23

dbalabka replied:
Yes, it is almost the same. Changed. Also, updated the file

oleg-andreyev commented on Tests/Functional/App/Basic/bundles.php:
```diff
@@ -0,0 +1,11 @@
+<?php
+/*
+ * This file is part of the OpCart software.
```
fix this

oleg-andreyev commented on Tests/Functional/App/YamlOnly/bundles.php:
```diff
@@ -0,0 +1,11 @@
+<?php
+/*
+ * This file is part of the OpCart software.
```
fix this

oleg-andreyev commented on Tests/Functional/App/config/bundles.php:
```diff
@@ -0,0 +1,19 @@
+<?php
+/*
+ * This file is part of the OpCart software.
```
fix this

oleg-andreyev commented on UPGRADE.md:
```diff
@@ -0,0 +1,14 @@
+Upgrade from 0.x to 1.x
```
this contradicts with 

```
Integration of JSON API with Symfony which supports JMS Serializer 2.0. 

 Note! Old version of JMS Serializer 1.0 support will end soon. 
Please use JsonApiBundle version [1.0](https://github.com/ecentria/JsonApiBundle/tree/1.0/) instead.
```

Also this file should be called CHANGELOG, but UPGRADE should have a brief guide of upgrade

dbalabka committed:
Updated the copyright headers. Some minor code adjustments

dbalabka committed:
Refactor DefaultAccessorStrategy to decorator

dbalabka committed:
Removed copy pasted YAML and Annotation drivers and cover it with functional tests

dbalabka committed:
Installed phpstan to catch most bugs (fixes #16)

dbalabka committed:
Minor fixes based on phpstan report in Drivers

dbalabka committed:
Created changelog and removed notes from UPGRADE.md

dbalabka committed:
Doc.block fixes and other minor issues

dbalabka committed:
Bumped version in composer.json

dbalabka committed:
Fixed PagerfantaHandler.
Removed redundant exclusion strategy.
Removed not needed jms_serializer decorator

dbalabka committed:
Enable PSR-4 in composer.json

dbalabka committed:
Update test to validate deserialization right after serialization.
Fixed bug with deserialization.

dbalabka commented:
Will address #8, #16

dbalabka committed:
Renamed package to ecentria/json-api-bundle and update the documentation

dbalabka committed:
Added missing copyright header and removed unused classes "use" declarations

videni commented:
@dbalabka , @oleg-andreyev when will this PR be merged?

dbalabka commented:
@videni thanks for reach us out. Right now we don’t have exact plan but PR pass all units and integrations test and looks ready do be merged. 
Are you going to use this in some project?

videni commented:
@dbalabka , Thanks for your quick reply,I am going to serve my project with JSON API,  It seems your PR can be used with the  Symfon 4  together seamlessly.hope it can be merged soon.

dbalabka commented:
@videni we will review the possibility to merge it. Unfortunately, it is not a high priority but we will do our best 😉
